# Week 11 

# Part 1: Tidying and Reshaping Data
## 0.0
Read the M11_Data.csv file from your GitHub repository into a
Jupyter Notebook WITHOUT removing any empty rows or columns from the content of the file.
The
content of the resulting dataframe should appear as follows:

In [99]:
import pandas as pd
import numpy as np
from numpy import nan as NA
df = pd.read_csv('https://raw.githubusercontent.com/JuJu-Ren/AIM-5001/main/M11_Data.csv', header=[0])
df

Month     Category      Caltex        Gulf       Mobil
0   Open   Engine Oil  140 :  000  199 :  000  141 :  000
1    NaN  GearBox Oil  198 :  000  132 :  000  121 :  000
2    Jan   Engine Oil  170 :  103  194 :  132  109 :  127
3    NaN  GearBox Oil  132 :  106  125 :  105  191 :  100
4    Feb   Engine Oil  112 :  133  138 :  113  171 :  101
5    NaN  GearBox Oil  193 :  148  199 :  119  134 :  127
6    Mar   Engine Oil  184 :  100  141 :  141  114 :  108
7    NaN  GearBox Oil  138 :  121  172 :  133  193 :  115
8    Apr   Engine Oil  149 :  150  117 :  118  117 :  118
9    NaN  GearBox Oil  185 :  125  191 :  133  119 :  121
10   May   Engine Oil  170 :  139  104 :  119  200 :  117
11   NaN  GearBox Oil  168 :  117  138 :  102  121 :  146
12   Jun   Engine Oil  159 :  129  170 :  138  169 :  105
13   NaN  GearBox Oil  107 :  129  195 :  141  141 :  112

## 1.1Combining and reshaping data in Pandas to tidy and transform/reshape
* We have many 'NaN' values in the 'Month' column as a result of the melt() operation

In [100]:
df["Month"].fillna(method='ffill', inplace = True)
df.head()

Month     Category      Caltex        Gulf       Mobil
0  Open   Engine Oil  140 :  000  199 :  000  141 :  000
1  Open  GearBox Oil  198 :  000  132 :  000  121 :  000
2   Jan   Engine Oil  170 :  103  194 :  132  109 :  127
3   Jan  GearBox Oil  132 :  106  125 :  105  191 :  100
4   Feb   Engine Oil  112 :  133  138 :  113  171 :  101

- We are defining the new name "suppliers" for the supplier names and "Value" for all the data value in each colomn

In [101]:
# Melting from wide to long format, first 2 columns are used to 
ids = ["Month","Category"]
# the remaining columns are treated as data values to be pivoted into long format
values = df.columns[2:]
df = pd.melt(frame=df, id_vars=ids, value_vars=values, var_name='Suppliers', value_name='Value')
df.head()

Month     Category Suppliers       Value
0  Open   Engine Oil    Caltex  140 :  000
1  Open  GearBox Oil    Caltex  198 :  000
2   Jan   Engine Oil    Caltex  170 :  103
3   Jan  GearBox Oil    Caltex  132 :  106
4   Feb   Engine Oil    Caltex  112 :  133

- split the purchased and consumed data into two columns

In [102]:
new = df["Value"].str.split(':', expand = True)
df["Purchased"],df["Consumed"]= new[0],new[1]
df.drop(columns = ["Value"], inplace = True)
df

Month     Category Suppliers Purchased Consumed
0   Open   Engine Oil    Caltex      140       000
1   Open  GearBox Oil    Caltex      198       000
2    Jan   Engine Oil    Caltex      170       103
3    Jan  GearBox Oil    Caltex      132       106
4    Feb   Engine Oil    Caltex      112       133
5    Feb  GearBox Oil    Caltex      193       148
6    Mar   Engine Oil    Caltex      184       100
7    Mar  GearBox Oil    Caltex      138       121
8    Apr   Engine Oil    Caltex      149       150
9    Apr  GearBox Oil    Caltex      185       125
10   May   Engine Oil    Caltex      170       139
11   May  GearBox Oil    Caltex      168       117
12   Jun   Engine Oil    Caltex      159       129
13   Jun  GearBox Oil    Caltex      107       129
14  Open   Engine Oil      Gulf      199       000
15  Open  GearBox Oil      Gulf      132       000
16   Jan   Engine Oil      Gulf      194       132
17   Jan  GearBox Oil      Gulf      125       105
18   Feb   Engine Oil      Gulf      138       113
19   Feb  GearBox Oil      Gulf      199       119
20   Mar   Engine Oil      Gulf      141       141
21   Mar  GearBox Oil      Gulf      172       133
22   Apr   Engine Oil      Gulf      117       118
23   Apr  GearBox Oil      Gulf      191       133
24   May   Engine Oil      Gulf      104       119
25   May  GearBox Oil      Gulf      138       102
26   Jun   Engine Oil      Gulf      170       138
27   Jun  GearBox Oil      Gulf      195       141
28  Open   Engine Oil     Mobil      141       000
29  Open  GearBox Oil     Mobil      121       000
30   Jan   Engine Oil     Mobil      109       127
31   Jan  GearBox Oil     Mobil      191       100
32   Feb   Engine Oil     Mobil      171       101
33   Feb  GearBox Oil     Mobil      134       127
34   Mar   Engine Oil     Mobil      114       108
35   Mar  GearBox Oil     Mobil      193       115
36   Apr   Engine Oil     Mobil      117       118
37   Apr  GearBox Oil     Mobil      119       121
38   May   Engine Oil     Mobil      200       117
39   May  GearBox Oil     Mobil      121       146
40   Jun   Engine Oil     Mobil      169       105
41   Jun  GearBox Oil     Mobil      141       112

#### The data is now in 'tidy' format:

- Each row is now comprised of a single observation 

## 1.2 Question 1
- What was the amount of oil remaining for each type/brand at the end of the chronological period?

In [116]:
df= df.astype({"Purchased":float, "Consumed":float})
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42 entries, 0 to 41
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Month      42 non-null     object 
 1   Category   42 non-null     object 
 2   Suppliers  42 non-null     object 
 3   Purchased  42 non-null     float64
 4   Consumed   42 non-null     float64
dtypes: float64(2), object(3)
memory usage: 1.8+ KB


In [118]:
grouped = df[["Purchased","Consumed"]].groupby(df['Category'])
grouped.describe()

Purchased                                                     \
                count        mean        std    min    25%    50%    75%   
Category                                                                   
Engine Oil       21.0  150.857143  30.969801  104.0  117.0  149.0  170.0   
GearBox Oil      21.0  156.809524  32.587757  107.0  132.0  141.0  191.0   

                   Consumed                                                   \
               max    count        mean        std  min    25%    50%    75%   
Category                                                                       
Engine Oil   200.0     21.0  104.333333  45.826121  0.0  103.0  118.0  132.0   
GearBox Oil  199.0     21.0  104.761905  45.806009  0.0  105.0  119.0  129.0   

                    
               max  
Category            
Engine Oil   150.0  
GearBox Oil  148.0

In [122]:
cf = grouped.sum()
cf

Purchased  Consumed
Category                        
Engine Oil      3168.0    2191.0
GearBox Oil     3293.0    2200.0

In [137]:
cf["Purchased"]-cf["Consumed"]

Category
Engine Oil      977.0
GearBox Oil    1093.0
dtype: float64

Here we can see, the amount of oil remaining for "Engine Oil" is 977 \
The amount of oil remaining for GearBox Oil is 1093\
## 1.2 Question 2
- What was the most consumed brand of oil across the two separate categories/types of oil?


In [146]:
grouped = df[["Consumed"]].groupby([df["Category"],df['Suppliers']])
grouped.sum()

Consumed
Category    Suppliers          
Engine Oil  Caltex        754.0
            Gulf          761.0
            Mobil         676.0
GearBox Oil Caltex        746.0
            Gulf          733.0
            Mobil         721.0

In the Engine Oil type, "Caltex" is the most consumed brand\
In the GearBox Oil type, "Caltex" also is the must consumed brand

## 1.3 
(15 Points) Finally, given your “tidy” long format structure, describe what, if any, changes you would make
to the visual presentation of the data if you were then asked to transform your “long” data back into a “wide”
format: would you mimic the structure of the graphic shown above? If not, how might you transform your
“long” data to “wide” format to make its “wide” presentation easier to understand and work with? Provide an
example of your recommendation and explain your rationale for preferring your specific structure.

In [149]:
df.shape,df.head()

((42, 5),
   Month     Category Suppliers  Purchased  Consumed
 0  Open   Engine Oil    Caltex      140.0       0.0
 1  Open  GearBox Oil    Caltex      198.0       0.0
 2   Jan   Engine Oil    Caltex      170.0     103.0
 3   Jan  GearBox Oil    Caltex      132.0     106.0
 4   Feb   Engine Oil    Caltex      112.0     133.0)

To switch the frame from long to wide, we choose to find the "longest" attribute, "Month" contains 7 values from "Open" to "June", so by pivoting the chart to a format with columns as Month, the dataframe can be effectively clean.

In [186]:
df_h=df.pivot_table(index = ["Category","Suppliers"],
              columns ="Month",
              values = ["Consumed","Purchased"], sort = True)
new_order = [('Purchased', 'Open'),('Purchased', 'Jan'),('Purchased', 'Feb'),('Purchased', 'Mar'),
             ('Purchased', 'Apr'),('Purchased', 'May'),('Purchased', 'Jun'),('Consumed', 'Open'),('Consumed', 'Jan'),
            ('Consumed', 'Feb'),('Consumed', 'Mar'),('Consumed', 'Apr'),( 'Consumed',  'May'),( 'Consumed',  'Jun')]
df_h.reindex(new_order, axis =1)

Purchased                                            \
Month                      Open    Jan    Feb    Mar    Apr    May    Jun   
Category    Suppliers                                                       
Engine Oil  Caltex        140.0  170.0  112.0  184.0  149.0  170.0  159.0   
            Gulf          199.0  194.0  138.0  141.0  117.0  104.0  170.0   
            Mobil         141.0  109.0  171.0  114.0  117.0  200.0  169.0   
GearBox Oil Caltex        198.0  132.0  193.0  138.0  185.0  168.0  107.0   
            Gulf          132.0  125.0  199.0  172.0  191.0  138.0  195.0   
            Mobil         121.0  191.0  134.0  193.0  119.0  121.0  141.0   

                      Consumed                                            
Month                     Open    Jan    Feb    Mar    Apr    May    Jun  
Category    Suppliers                                                     
Engine Oil  Caltex         0.0  103.0  133.0  100.0  150.0  139.0  129.0  
            Gulf           0.0  132.0  113.0  141.0  118.0  119.0  138.0  
            Mobil          0.0  127.0  101.0  108.0  118.0  117.0  105.0  
GearBox Oil Caltex         0.0  106.0  148.0  121.0  125.0  117.0  129.0  
            Gulf           0.0  105.0  119.0  133.0  133.0  102.0  141.0  
            Mobil          0.0  100.0  127.0  115.0  121.0  146.0  112.0

# Part 2: Using Your GroupBy and Data Aggregation Skills
- Load the data from file  briges.data.version1.csv  in GitHub repository

#### Attribute Information:

The type field state whether a property is continuous/integer (c) or nominal (n).

For properties with c,n type, the range of continuous numbers is given first and the possible values of the nominal follow the semi-colon.

Name / Type / Possible values / Comments

1. IDENTIF / -- / -- / identifier of the examples
2. RIVER / n / A, M, O / --
3. LOCATION / n / 1 to 52 / --
4. ERECTED / c,n / 1818-1986 ; CRAFTS, EMERGING, MATURE, MODERN / --
5. PURPOSE / n / WALK, AQUEDUCT, RR, HIGHWAY / --
6. LENGTH / c,n / 804-4558 ; SHORT, MEDIUM, LONG / --
7. LANES / c,n / 1, 2, 4, 6 ; 1, 2, 4, 6 / --
8. CLEAR-G / n / N, G / --
9. T-OR-D / n / THROUGH, DECK / --
10. MATERIAL / n / WOOD, IRON, STEEL / --
11. SPAN / n / SHORT, MEDUIM, LONG / --
12. REL-L / n / S, S-F, F / --
13. TYPE / n / WOOD, SUSPEN, SIMPLE-T, ARCH, CANTILEV, CONT-T / --

In [191]:
data = pd.read_csv('https://raw.githubusercontent.com/JuJu-Ren/AIM-5001/main/bridges.data.version1.csv', header=None)
data.columns =["ID","River","Location","Erected","Purpose","Length","Lanes",
               "Clear-G","T/D","Material","Span","Rel-l","Type"]
data

ID River Location  Erected   Purpose Length Lanes Clear-G      T/D  \
0      E1     M        3     1818   HIGHWAY      ?     2       N  THROUGH   
1      E2     A       25     1819   HIGHWAY   1037     2       N  THROUGH   
2      E3     A       39     1829  AQUEDUCT      ?     1       N  THROUGH   
3      E5     A       29     1837   HIGHWAY   1000     2       N  THROUGH   
4      E6     M       23     1838   HIGHWAY      ?     2       N  THROUGH   
..    ...   ...      ...      ...       ...    ...   ...     ...      ...   
103   E84     A       24     1969   HIGHWAY    870     6       G  THROUGH   
104   E91     O       44     1975   HIGHWAY   3756     6       G  THROUGH   
105   E90     M        7     1978   HIGHWAY    950     6       G  THROUGH   
106  E100     O       43     1982   HIGHWAY      ?     ?       G        ?   
107  E109     A       28     1986   HIGHWAY      ?     ?       G        ?   

    Material    Span Rel-l  Type  
0       WOOD   SHORT     S  WOOD  
1       WOOD   SHORT     S  WOOD  
2       WOOD       ?     S  WOOD  
3       WOOD   SHORT     S  WOOD  
4       WOOD       ?     S  WOOD  
..       ...     ...   ...   ...  
103    STEEL  MEDIUM     F  ARCH  
104    STEEL    LONG     F  ARCH  
105    STEEL    LONG     F  ARCH  
106        ?       ?     F     ?  
107        ?       ?     F     ?  

[108 rows x 13 columns]

## 2.1
how many bridges of each ‘Purpose’/’Material’ grouping within the data set have been constructed over each of the rivers listed in the data set.

In [235]:
grouped = data["Material"].groupby([data["River"],data['Purpose'],data["Material"]]).agg(['count'])
grouped.rename(columns ={'count':'How Many?'})

How Many?
River Purpose  Material           
A     AQUEDUCT IRON              1
               WOOD              3
      HIGHWAY  ?                 1
               IRON              2
               STEEL            21
               WOOD              8
      RR       IRON              1
               STEEL             9
               WOOD              2
      WALK     STEEL             1
M     HIGHWAY  IRON              4
               STEEL            19
               WOOD              3
      RR       IRON              2
               STEEL            13
O     HIGHWAY  ?                 1
               IRON              1
               STEEL             9
      RR       STEEL             4
Y     HIGHWAY  STEEL             2
      RR       STEEL             1

## 2.2
show the average length for each ‘Purpose’/’Material’ bridge grouping within the data set

In [254]:
# cleaning up the ‘Length’ column 
data["Length"] = data["Length"].replace('?', np.nan)
data =  data.astype({'Length':float})
# creating the report

# using "mean" to represent "average length", group by purpose, mertial in the requested order
grouped = data["Length"].groupby([data['Purpose'],data["Material"]]).agg(['mean'])
# rename the column
grouped.rename(columns ={'mean':'Average Length'})

Average Length
Purpose  Material                
AQUEDUCT IRON         1000.000000
         WOOD         1092.000000
HIGHWAY  ?                    NaN
         IRON         1216.666667
         STEEL        1557.804348
         WOOD         1053.375000
RR       IRON         1100.000000
         STEEL        1946.850000
         WOOD                 NaN
WALK     STEEL                NaN

## 2.3 
report that shows the average length, count, minimum length, and maximum length of bridges built during 4 equal length time periods (1818 – 1860; 1860-
1902; 1902-1944; 1944-1986).

In [277]:
# check the "Erected"category, divide it with 4 equal length, set it as period
period = pd.cut(data['Erected'],4)
period[::30]

0     (1817.832, 1860.0]
30      (1860.0, 1902.0]
60      (1902.0, 1944.0]
90      (1944.0, 1986.0]
Name: Erected, dtype: category
Categories (4, interval[float64, right]): [(1817.832, 1860.0] < (1860.0, 1902.0] < (1902.0, 1944.0] < (1944.0, 1986.0]]

In [279]:
def get_stats(group):
    return {'Average Length': group.mean(),'Count': group.count(), 'Max Length': group.max(),
            'Min Length': group.min()}
# short "Erected" with 4 cuts
grouped = data["Length"].groupby(period)
# change the names in multiple levels column 
grouped.apply(get_stats).unstack()

Average Length  Count  Max Length  Min Length
Erected                                                          
(1817.832, 1860.0]     1094.625000    8.0      1500.0       990.0
(1860.0, 1902.0]       1603.347826   23.0      4558.0      1000.0
(1902.0, 1944.0]       1676.181818   33.0      3000.0       860.0
(1944.0, 1986.0]       1530.411765   17.0      3756.0       804.0